Mathlib/Topology/Constructible.lean

```diff
@@ -0,0 +1,303 @@
+    IsConstructible (f ⁻¹' s) := by
+  induction hs using IsConstructible.empty_union_induction with
+  | open_retrocompact U hUopen hUcomp =>
+    exact (hf _ hUcomp).isConstructible <| hUopen.preimage hfcont
+  | union s hs t ht hs' ht' => rw [preimage_union]; exact hs'.union ht'
+  | compl s hs hs' => rw [preimage_compl]; exact hs'.compl
+
+@[stacks 005J]
+lemma IsConstructible.preimage_of_isOpenEmbedding {s : Set β} (hf : IsOpenEmbedding f)
+    (hs : IsConstructible s) : IsConstructible (f ⁻¹' s) :=
+  hs.preimage hf.continuous fun _t ht ↦ ht.preimage_of_isOpenEmbedding hf
+
+@[stacks 09YE]
+lemma IsConstructible.preimage_of_isClosedEmbedding {s : Set β} (hf : IsClosedEmbedding f)
+    (hf' : IsCompact (range f)ᶜ) (hs : IsConstructible s) : IsConstructible (f ⁻¹' s) :=
+  hs.preimage hf.continuous fun _t ht ↦ ht.preimage_of_isClosedEmbedding hf hf'
+
+@[stacks 09YD]
+lemma IsConstructible.image_of_isOpenEmbedding (hfopen : IsOpenEmbedding f)
+    (hfcomp : IsRetroCompact (range f)) (hs : IsConstructible s) : IsConstructible (f '' s) := by
+  induction hs using IsConstructible.empty_union_induction with
+  | open_retrocompact U hUopen hUcomp =>
+    exact (hUcomp.image_of_isEmbedding hfopen.isEmbedding hfcomp).isConstructible <|
+      hfopen.isOpenMap _ hUopen
+  | union s hs t ht hs' ht' => rw [image_union]; exact hs'.union ht'
+  | compl s hs hs' =>
+    rw [← range_diff_image hfopen.injective]
+    exact (hfcomp.isConstructible hfopen.isOpen_range).sdiff hs'
+
+@[stacks 09YG]
+lemma IsConstructible.image_of_isClosedEmbedding (hf : IsClosedEmbedding f)
+    (hfcomp : IsRetroCompact (range f)ᶜ) (hs : IsConstructible s) : IsConstructible (f '' s) := by
+  induction hs using IsConstructible.empty_union_induction with
+  | open_retrocompact U hUopen hUcomp =>
+    have hfU : IsOpen (f '' U ∪ (range f)ᶜ) := by
+      simpa [← range_diff_image hf.injective, sdiff_eq, compl_inter, union_comm]
+        using (hf.isClosedMap _ hUopen.isClosed_compl).isOpen_compl
+    suffices h : IsRetroCompact (f '' U ∪ (range f)ᶜ) by
+      simpa [union_inter_distrib_right, inter_eq_left.2 (image_subset_range ..)]
+        using (h.isConstructible hfU).sdiff (hfcomp.isConstructible hf.isClosed_range.isOpen_compl)
+    rintro V hVcomp hVopen
+    rw [union_inter_distrib_right, ← image_inter_preimage]
+    exact ((hUcomp (hf.isCompact_preimage hVcomp) (hVopen.preimage hf.continuous)).image
+      hf.continuous).union <| hfcomp hVcomp hVopen
+  | union s hs t ht hs' ht' => rw [image_union]; exact hs'.union ht'
+  | compl s hs hs' =>
+    rw [← range_diff_image hf.injective]
+    exact (hfcomp.isConstructible hf.isClosed_range.isOpen_compl).of_compl.sdiff hs'
+
+section CompactSpace
+variable [CompactSpace α] {P : ∀ s : Set α, IsConstructible s → Prop} {b : ι → Set α}
+
+lemma IsRetroCompact.isCompact (hs : IsRetroCompact s) : IsCompact s := by
+  simpa using hs CompactSpace.isCompact_univ
+
+lemma TopologicalSpace.IsTopologicalBasis.isRetroCompact_iff_isCompact
```

Comment:
Why working with a parameterized topological basis `b` and not a `Set (Set alpha)`?

Reply:
(that should remove some `lift` subtleties below)

Reply:
Waiting for @erdOne's input here

Reply:
I responded to Yael by DM and forgot to reply here.
Yes I agree that it is better to use `Set (Set alpha)`

Reply:
Note that some things got better, some things got worse. Likely, the best of both worlds would be to have two predicates for topological bases: one for set families, one for indexed families